src/compiler/checker.ts

```diff
@@ -17651,12 +17652,51 @@ namespace ts {
             }
         }
 
+        function getOptionalTypeRecursive(type: Type) {
+            if ((type as ResolvedType).properties) {
+                const rType = type as ResolvedType;
+                const propCount = rType.properties.length;
+
+                if (rType.flags & TypeFlags.Object) {
+                  for (let i = 0; i < propCount; i++) {
+                    const prop = rType.properties[i] as SymbolLinks;
+                    let innerType;
+                    if (prop.bindingElement && prop.bindingElement.initializer) {
+                        innerType = getOptionalTypeRecursive(prop.type);
+                        innerType = getOptionalType(innerType);
+                    }
+                    else if (prop.type.symbol) {
+                        innerType = getTypeOfParameter(prop.type.symbol);
+                    }
+                    else {
+                        innerType = prop.type;
+                    }
+                    prop.type = innerType;
+                    rType.properties[i] = prop as Symbol;
+                  }
+                }
+
+                if (type.flags & TypeFlags.Union || type.flags & TypeFlags.Intersection) {
+                  for (let i = 0; i < propCount; i++) {
+                    const prop = rType.properties[i] as SymbolLinks;

```

Comment:
This only works if the union has already had its properties resolved. `getPropertiesOfType` does the right thing.